### PR TITLE
Port BasisRotation to Operator2

### PR DESCRIPTION
**Pull request by comp-phys-marc**
**Context:** We would like to try porting a simple template to `Operator2`.

**Description of the Change:** Ports `BasisRotation` to `Operator2`.

**Benefits:** We get an idea of the effort involved. Also helps unlock our goals.

**Possible Drawbacks:** It is a very simple example.

**Related GitHub Issues:** [sc-125433]

### CODE REVIEW

File: tests/capture/test_templates.py [hunk 1]
```diff
@@ -483,33 +484,26 @@ def qfunc(U, O):
...
 
         # Actually test primitive bind
-        jaxpr = jax.make_jaxpr(qfunc)(wires, mat)
+        jaxpr = jax.make_jaxpr(qfunc)(mat, wires)
 
         assert len(jaxpr.eqns) == 1
 
         eqn = jaxpr.eqns[0]
-        assert eqn.primitive == qp.BasisRotation._primitive
+        assert eqn.primitive == operator_p
```
Comment (andrijapau) [inline tests/capture/test_templates.py]: You should also check if its the right op_cls

Comment (comp-phys-marc) [reply tests/capture/test_templates.py]: Thanks

Comment (comp-phys-marc) [reply tests/capture/test_templates.py]: https://github.com/PennyLaneAI/pennylane/pull/9896/commits/f8d7dc00d706639a39b41e307c9aee4b3b71249f

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 2]
```diff
@@ -289,24 +290,16 @@ class BasisRotation(Operation):
...
-    resource_keys = {"dim", "is_real"}
-
-    @classmethod
-    def _primitive_bind_call(cls, wires, unitary_matrix, check=False):
-        # pylint: disable=arguments-differ
-        if cls._primitive is None:
-            # guard against this being called when primitive is not defined.
-            return type.__call__(cls, wires, unitary_matrix, check=check)  # pragma: no cover
+    wires_argnames = ("wires",)
+    static_argnames = ("check",)
```
Comment (astralcai) [inline pennylane/templates/subroutines/qchem/basis_rotation.py]: `check` being a `boolean`, I think it's compilable right?

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: oh that's a good point

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 3]
```diff
@@ -325,17 +318,14 @@ def __init__(self, wires, unitary_matrix, check=False):
 
         super().__init__(unitary_matrix, wires=wires)
```
Comment (astralcai) [inline pennylane/templates/subroutines/qchem/basis_rotation.py:319]: In this case, I don't think `check` should be in any argname at all. It's not a property of the operator, it's simply a toggle for whether the `__init__` should check the input. Two `BasisRotations` with identical `unitary_matrix` and `wires` are considered identical, even if one of them is instantiated with `check=False` and the other one with `check=True`.

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:319]: But it is an arg, so we will raise an error if it is not an argname

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 4]
```diff
@@ -398,7 +388,11 @@ def compute_decomposition(
         return op_list
 
 
-def _basis_rotation_decomp_resources(dim, is_real):
+# pylint: disable=unused-argument
+def _basis_rotation_decomp_resources(wires, unitary_matrix, check=False):
```
Comment (astralcai) [inline pennylane/templates/subroutines/qchem/basis_rotation.py]:
```suggestion
def _basis_rotation_decomp_resources(unitary_matrix, wires):
```

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 5]
```diff
@@ -398,7 +388,11 @@ def compute_decomposition(
...
+    dim = math.shape(unitary_matrix)[0]
+    is_real = math.is_real_obj_or_close(unitary_matrix)
```
Comment (astralcai) [inline pennylane/templates/subroutines/qchem/basis_rotation.py]: This won't work anymore, because `unitary_matrix` here is gonna be an `AbstractArray((2,), dtype='complex128')`

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: oh right

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: actually we should still be able to get the dim

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 6]
```diff
@@ -409,13 +403,15 @@ def _basis_rotation_decomp_resources(dim, is_real):
...
+def _basis_rotation_decomp(unitary_matrix, wires, **__):
 
     if isinstance(wires, Wires):
         wires = wires.labels
 
     if _qjit_or_capture():
-        matrix, wires = math.array(matrix, like="jax"), math.array(wires, like="jax")
+        unitary_matrix, wires = math.array(unitary_matrix, like="jax"), math.array(
+            wires, like="jax"
+        )
```
Comment (astralcai) [inline pennylane/templates/subroutines/qchem/basis_rotation.py:406]: I wonder if things like this is going to be required anymore.

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 5 above)
Comment (astralcai) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: Yeah the comment is on the `is_real_obj_or_close` line. The `shape` should work? You might want to test it.

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 6 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:406]: hm. I don't see why we would be able to get rid of it yet?

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 3 above)
Comment (astralcai) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:319]: Let's bring this up in code discussion today.

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 6 above)
Comment (astralcai) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:406]: Because Catalyst should have done it for us?

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:406]: oh I see

Comment (astralcai) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:406]: Can you remove this to see if anything fails?

Comment (astralcai) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:406]: I think `DecomposeInterpreter` still depends on this behaviour.

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 5 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]:
```
>>> math.shape(Wire[1])
(1,)
```

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 6 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:406]: ok I'll leave it then

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 7]
```diff
@@ -398,24 +388,26 @@ def compute_decomposition(
...
 
+    se_count = dim * (dim - 1) / 2
     ps_count = dim + se_count
     return {PhaseShift: ps_count, SingleExcitation: se_count}
 
 
 # Not exact because PhaseShift(s) might be skipped
 @register_resources(_basis_rotation_decomp_resources, exact=False)
-def _basis_rotation_decomp(matrix, wires: WiresLike, **__):
+def _basis_rotation_decomp(unitary_matrix, wires, **__):
```
Comment (astralcai) [inline pennylane/templates/subroutines/qchem/basis_rotation.py:398]: Just making sure, is this decomposition rule tested anywhere?

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:398]: yes it is tested

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 8]
```diff
@@ -289,24 +290,16 @@ class BasisRotation(Operation):
...
 
-        return cls._primitive.bind(*wires, unitary_matrix, check=check)
-
-    @classmethod
-    def _unflatten(cls, data, metadata):
-        return cls(wires=metadata[0], unitary_matrix=data[0])
+    grad_method = None
 
-    def __init__(self, wires, unitary_matrix, check=False):
+    def __init__(self, unitary_matrix, wires, check=False):
```
Comment (andrijapau) [inline pennylane/templates/subroutines/qchem/basis_rotation.py:302]: Why did we switch the order?

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 3 above)
Comment (andrijapau) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:319]: What was the conclusion here?

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 9]
```diff
@@ -325,17 +318,14 @@ def __init__(self, wires, unitary_matrix, check=False):
...
-            "dim": math.shape(self.data[0])[0],
-            "is_real": math.is_real_obj_or_close(self.data[0]),
-        }
-
     @property
     def num_params(self):
         return 1
 
+    @property
+    def ndim_params(self):
```
Comment (andrijapau) [inline pennylane/templates/subroutines/qchem/basis_rotation.py]: Why did you need to override this?

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 8 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:302]: because dynamic args come before wires args generally?

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 9 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: I think it wasn't defined

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 3 above)
Comment (astralcai) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:319]: We leave it in `compilable_args`

**Review by astralcai — APPROVED** [review]

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 8 above)
Comment (andrijapau) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:302]: True. Will be a breaking change though right? I guess we're fine with that?

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 9 above)
Comment (andrijapau) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: It should be defined by `Operator2` now, right?

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: Operator2 just defaults to defining it as -1. The individual operators are supposed to define it for themselves.

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 8 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:302]: Yes we are

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 9 above)
Comment (andrijapau) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: I don't think that's the case - we have logic that calculates it. I tried locally by deleting this addition and I got,
```python
import pennylane as qp
op = qp.BasisRotation([[0,1], [1,0]], wires=[0, 1])
>>> print(op.ndim_params)
(2,)
```

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 8 above)
Comment (andrijapau) [reply pennylane/templates/subroutines/qchem/basis_rotation.py:302]: confirmed with @isaacdevlugt . all good!

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 9 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: ok, deleted

Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: https://github.com/PennyLaneAI/pennylane/pull/9896/commits/2989387fe6e3afde8b26d18d1d624c65428bdb08

**Review by andrijapau — APPROVED** [review]
LGTM!!!! 🚀 🚀
